Added documentation for quarkus.package.type=native-sources (#18589)
quarkusio/quarkus: docs/src/main/asciidoc/building-native-image.adoc

It contains the required packages to run a native executable and is only **9Mb**.
Just add your application on top of this image, and you will get a tiny container image.

=== Separating Java- and native-image compilation

> In certain circumstances, you may want to separate compilation of the Java and native-image.

Reviewer: ```suggestion
In certain circumstances, you may want to build the native image in a separate step.
```

> For example, in a CI/CD pipeline, you may want to have one step to generate the JAR file and another step to compile this JAR file into the native executable.

Reviewer: ```suggestion
For example, in a CI/CD pipeline, you may want to have one step to generate the source that will be used for the native image generation and another step to use these sources to actually build the native executable.
```

For this use case, you can set the `quarkus.package.type=native-sources`.

> This will execute the java compilation as if you would have started native compilation (`-Pnative`), but stop before triggering the actual call to GraalVM's `native-image`.

Reviewer: ```suggestion
This will execute the java compilation as if you would have started native compilation (`-Pnative`), but stops before triggering the actual call to GraalVM's `native-image`.
```

[source,bash]
---
$ ./mvnw clean package -Dquarkus.package.type=native-sources
---

After compilation has finished, you find the build artifact in `target/native-sources`:

[source,bash]
---
$ cd target/native-sources
$ ls
native-image.args  getting-started-1.0.0-SNAPSHOT-runner.jar
---

You see that, beside the expected JAR, a file `native-image.args` was created.
This file holds all parameters (including the name of the JAR to compile) to pass along to GraalVM's `native-image` command.
If you have GraalVM installed, you can start the native compilation by executing:

[source,bash]
---
$ cd target/native-source
$ native-image $(cat native-image.args)
...
$ ls
native-image.args
getting-started-1.0.0-SNAPSHOT-runner
getting-started-1.0.0-SNAPSHOT-runner.build_artifacts.txt
getting-started-1.0.0-SNAPSHOT-runner.jar
---

The process for Gradle is analogous.

> Of course, you can also run the compilation in a container:

Reviewer: ```suggestion
Running the build process in a container is also possible:
```

cd target/native-image
docker run \
  -it \
  --rm \
  --v $(pwd):/work <1>
  -w /work <2>
  --entrypoint bin/sh \
  quay.io/quarkus/ubi-quarkus-native-image:21.1.0-java11 \ <3>
  -c "native-image $(cat native-image.args) -J-Xmx4g" <4>
---

<1> Mount the host's directory `target/native-image` to the container's `/work`. Thus, the generated binary will also be written to this directory.
<2> Switch the working directory to `/work`, which we have mounted in <1>.
<3> Use the `quay.io/quarkus/ubi-quarkus-native-image:21.1.0-java11` docker image introduced in <<#multistage-docker,Using a multi-stage Docker build>> to build the native image.
<4> Call `native-image` with the content of file `native-image.args` as arguments. We also supply an additional argument to limit the process's maximum memory limitation to 4 Gigabyte.

In a CI/CD setup, we would:

1. register the output of the step executing `./mvnw ...` command (i.e. directory `target/native-image`) as build artifact,
2. require this artifact in the step executing the `native-image ...` command, and
3. register the output of the step executing the `native-image ...` command (i.e. files matching `target/*runner`) as build artifact.

> The environment executing 1. only needs access to a Java- and a Maven- (or Gradle)-installation, while the environment executing 3. only needs access to a GraalVM installation (including the `native-image` feature).

Reviewer: ```suggestion
The environment executing `1` only needs access to a Java- and a Maven- (or Gradle)-installation, while the environment executing `3` only needs access to a GraalVM installation (including the `native-image` feature).
```

The generated binary can then be used to create a container image, as explained above.

[WARNING]
===

> If you are running under Windows, please keep in mind that the binary was created within a Linux docker container.
> Hence, the binary will not be executable under Windows.

Reviewer: ```suggestion
If you are running on a Windows machine, please keep in mind that the binary was created within a Linux docker container.
Hence, the binary will not be executable on the host Windows machine.
```
